ports/stm32/modmachine.c

```diff
@@ -277,7 +278,7 @@ STATIC NORETURN mp_obj_t machine_bootloader(size_t n_args, const mp_obj_t *args)
     #if MICROPY_HW_USES_BOOTLOADER
     if (n_args == 0 || !mp_obj_is_true(args[0])) {
         // By default, with no args given, we enter the custom bootloader (mboot)
-        powerctrl_enter_bootloader(0x70ad0000, MBOOT_VTOR);
+        MICROPY_BOARD_ENTER_BOOTLOADER(0x70ad0000, MBOOT_VTOR);
```

Comment:
Could you give an example of how a board would override this macro?  The arguments passed here are very specific (they mean jump to MBOOT_VTOR with r0=0x70ad0000) and probably don't have much meaning if the board overrides this macro.

Maybe instead a board just needs to override `machine_bootloader()` itself?  It really depends on what you are intending to do here.

Reply:
> Could you give an example of how a board would override this macro? The arguments passed here are very specific (they mean jump to MBOOT_VTOR with r0=0x70ad0000) and probably don't have much meaning if the board overrides this macro.

How about using variable args macro ?
```c
// boardctrl.h
#ifndef MICROPY_BOARD_ENTER_BOOTLOADER
MICROPY_BOARD_BOOTLOADER_ARGS   0x70ad0000, MBOOT_VTOR
MICROPY_BOARD_ENTER_BOOTLOADER(...) powerctrl_enter_bootloader(__VA_ARGS__)
#endif

// To override in some board config file:
MICROPY_BOARD_BOOTLOADER_ARGS
MICROPY_BOARD_ENTER_BOOTLOADER(...) myboard_enter_bootloader(__VA_ARGS__)

//Finally in modmachine.c call
MICROPY_BOARD_ENTER_BOOTLOADER(MICROPY_BOARD_BOOTLOADER_ARGS);
```

My end goal is to provide a consistent way (for all ports) to call custom bootloader entry code like this https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/board_init.c#L127

Reply:
> Maybe instead a board just needs to override machine_bootloader()

Yes that should work too I guess, and it's common across ports. The last time discussed this, it wasn't an option, which is why I sent this PR this way. The option was to override `branch_to_bootloader` .

Reply:
OK, so it looks like the scope is for a board to implement a custom way of entering the bootloader (yes that's pretty obvious!).  Currently it supports the following configurations to get into the bootloader:
- boards that use mboot, or just plain ST bootloader
- boards that jump to the bootloader or do a reset to get into the bootloader

The mboot case parses arguments to `machine.bootloader()` and has a fall-back to the ST bootloader.

What the Portenta H7 needs is none of the above, it needs to reset into the bootloader with an RTC backup register set to a special value.  So probably the best way to do this is just make the body of machine bootloader configurable, eg:
```c
STATIC NORETURN mp_obj_t machine_bootloader(size_t n_args, const mp_obj_t *args) {
    #if MICROPY_HW_ENABLE_USB
    pyb_usb_dev_deinit();
    #endif
    #if MICROPY_HW_ENABLE_STORAGE
    storage_flush();
    #endif
    
    __disable_irq();
    
    MICROPY_BOARD_ENTER_BOOTLOADER(n_args, arg);

    #if defined(STM32F7) || defined(STM32H7)
    powerctrl_enter_bootloader(0, 0x1ff00000);
    #else
    powerctrl_enter_bootloader(0, 0x00000000);
    #endif

    while (1) {
        ;
    }
}
```
Then the default implementation is what is already there.  If `MICROPY_BOARD_ENTER_BOOTLOADER()` returns then it'll fall back to the ST bootloader.

Reply:
> So probably the best way to do this is just make the body of machine bootloader configurable, eg:

Okay but in this case `machine_bootloader` should be exposed, so it's more convenient to call it from CDC drivers for example.